odml/tools/fuzzy_finder.py

```diff
@@ -0,0 +1,203 @@
+import pprint
+import time
+
+from odml.tools.query_creator import QueryCreator, QueryParser, QueryParser2
+
+
+class FuzzyFinder(object):
+    """
+    FuzzyFinder tool for querying graph through 'fuzzy' queries. 
+    If the user do not know exact attributes and structure of the odML data model,
+    the finder executes multiple queries to better match the parameters and returns sets of triples.
+    """
+    def __init__(self, graph=None, q_params=None):
+        self.graph = graph
+        self.q_params = q_params
+        self.prepared_queries_list = []
+        self._subsets = []
+
+    def find(self, graph=None, q_str=None, q_params=None):
+        """
+        Apply set of queries to the graph and returns info that was retrieved from queries.
+        
+        :param graph: graph object
+        :param q_params: dictionary object with set of parameters for a query
+                         Example: {'Sec': [('name', 'Stimulus')],
+                                   'Doc': [('author', 'D. N. Adams')],
+                                   'Prop': [('name', 'Contrast'), ('value':[20]), ('unit':'%')]}
+        :param q_str: query string which used in QueryCreator class
+                      Example: doc(author:D. N. Adams) section(name:Stimulus) prop(name:Contrast, value:20, unit:%)
+        :return: string which contains set of triples
+        """
+        self._validate_find_input_attributes(graph, q_str, q_params, QueryParser())
+
+        self._generate_parameters_subsets(self._generate_parameters_pairs())
+
+        return self._output_query_results()
+
+    def find_fuzzy(self, graph=None, q_str=None, q_params=None):
+        """
+        :param q_params: dictionary object with set of parameters for a query
+               Example: {'Sec': ['name', 'type'],
+                         'Doc': ['author'],
+                         'Search': ['Stimulus', 'Contrast']}
+        :param q_str: query string which used in QueryCreator class
+               Example: "select sec(name) prop(type) where Stimulus, Contrast"
+        :param graph: graph object
+        :return: string which contains set of triples
+        """
+        self._validate_find_input_attributes(graph, q_str, q_params, QueryParser2())
+
+        self._generate_parameters_subsets(self._generate_parameters_pairs_fuzzy())
+
+        return self._output_query_results()
+
+    def _validate_find_input_attributes(self, graph, q_str, q_params, q_parser):
+        if not graph:
+            self.graph = graph
+        if q_str and q_params:
+            raise ValueError("Please pass query parameters only as string or dict object")
+
+        if q_str:
```

Comment:
With this, `find_fuzzy` will always use only the query string, if both query string and query parameters are provided. Query parameters will only be used, if no query string is provided. Is this the intended behavior? From the docstring this behavior is not apparent.

Reply:
The idea is that we can accept either string with query or directly params but not both and not nothing (logical XOR). In my input validation I raise errors exactly to this logic.
So yes, I will use query params if no query string provided.